### PR TITLE
More `optional_type` conformance tests

### PR DESCRIPTION
**Pull request by jcking**

### CODE REVIEW

File: tests/simple/testdata/optionals.textproto [hunk 1]
```diff
@@ -234,4 +234,113 @@ section: {
...
+    name: "optional_empty_struct_optindex_hasValue"
+    container: "google.api.expr.test.v1.proto2"
+    expr: "optional.of(TestAllTypes{}).?repeated_string.hasValue()"
+    value: {
+      bool_value: false
+    }
+  }
+  test {
+    name: "optional_none_optselect_hasValue"
+    expr: "optional.none().?repeated_string.hasValue()"
```
Comment (TristonianJones) [inline tests/simple/testdata/optionals.textproto:262]: This works? I would think the type-checker would complain about this

File: tests/simple/testdata/optionals.textproto [hunk 2]
```diff
@@ -234,4 +234,113 @@ section: {
...
+        values {
+          string_value: "foo"
+        }
+      }
+    }
+  }
+  test {
+    name: "optional_struct_optindex_value"
+    container: "google.api.expr.test.v1.proto2"
+    expr: "optional.of(TestAllTypes{repeated_string: ['foo']}).?repeated_string.value()"
```
Comment (TristonianJones) [inline tests/simple/testdata/optionals.textproto:282]: Can you add a variant of this that attempts to get the element at `repeated_string[0].value()` as well?

Comment (jcking) [reply tests/simple/testdata/optionals.textproto:282]: Done.

File: tests/simple/testdata/optionals.textproto (hunk 1 above)
Comment (jcking) [reply tests/simple/testdata/optionals.textproto:262]: Yeah, `optional.none()` should be `optional_type<dyn>`.

**Review by TristonianJones — APPROVED** [review]

**Pull request merged**